content/blog/2019-09-25-making-function-calls-work.md

```diff
@@ -0,0 +1,299 @@
+The provided interpreter already supports the evaluation of functions. All we have to do is add a *call* instruction to the language grammar and link that to the interpreter. To limit function scope, we create a new environment (a map from variable names to values) every time a function is called.
+
+Below is a Bril program that demonstrates this functionality.
+
+    # This program prints out 100 and exits. Nothing too exciting here.
+    main {
+      call func;
+    }
+    
+    func {
+      v0: int = const 100;
+      print v0;
+    }
+
+Since the interpreter scans function definitions before executing anything, functions can be defined in any order. The above program can be rewritten as:
+
+    func {
+      v0: int = const 100;
+      print v0;
+    }
+
+    main {
+      call func;
+    }
+
+
+#### Adding function parameters
+To add support for function parameters, we first update the grammars for function definitions and calls to take whitespace-delimited lists of variable names. The interpreter's `call` operation handler first extracts the values for all arguments from the current environment map. It then pre-populates the callee's environment with the function parameters mapped to these values. This new environment is then used when evaluating the called function.
+
+Below is a Bril program that demonstrates the use of function parameters.
+
+    # This program also prints out 100 and exits.
+    main {
+      call print_double 50;
+    }
+    
+    print_double x {
+      v0: int = add x x;
+      print v0;
+    }
+
+#### Return types
+Without the ability to return data, our functions are not very useful. To add support for return types, we first update the grammar for function definitions to optionally take `: type` at the end of the header, and update the rule for `ret` to optionally take a variable name to return. We also overload the `call` operation to be both an effect operation and a value operation, since functions that do not return anything will be effect operations and those that do will be value operations.
+
+To pass a return value back to the caller, we use the following approach. When we handle a `ret` operation, if there is a return value, we add a special variable `_ ret` to the environment and map it to the return value. Note that this will never collide with any existing variables as there is space in the name. Then, if the caller is expecting a return value (which it will in instructions of the form `v: type = call func_name`), it can check for the existence of `_ ret` in the callee's environment. We do some basic type-checking here, comparing return types and variable types, to make the Bril programmer's life easier.
+
+Below is an example program that demonstrates this functionality.
+
+    # Once again, we print 100.
+    main {
+      v0: int = call get_hundred;
+      print v0;
+    }
+    
+    get_hundred: int {  # The ": int" is required.
+      v0: int = const 100;
+      ret v0;
+    }
```

Comment:
As above, it would be nice to discuss how this extension shows up in the abstract syntax.